docs/language-server/protocol-language-server.md

```diff
@@ -1124,8 +1126,43 @@ A representation of a change to a text file at a given position.
 
 ```typescript
 interface TextEdit {
+  /** The range of text in a text file. */
   range: Range;
+
+  /** The change to a text file. */
   text: String;
+
+  /** The update type. When not set defaults to `FileEditKind`. */
+  kind?: EditKind;
+}
+```
+
+### EditKind
+
+Represents the type of an update enabling runtime optimizations.
+
+- `FileEditKind` represents a plane file update. After applying the text edit,
+  the program will be recompiled and re-executed.
+- `MetadataEditKind` represents the metadata update. Metadata update does not
+  change the program and thus allows engine to skip inane re-executions.
+- `ValueUpdateKind` represents an update of a node value. In this case the
+  engine can skip the compilation phase, and re-execute the program with a new
+  value.
+
+#### Format
+
+```typescript
+type EditKind = FileEditKind | MetadataEditKind | ValueUpdateKind;
+
+interface FileEditKind {}
+
+interface MetadataEditKind {}
+
+interface ValueUpdateKind {
+  /** The identifier of altered node. */
+  expressionId: ExpressionId;
+  /** The new value of altered node. */
+  value: string;
```

Comment:
So far the value is taken from `TextEdit.text`. Can you describe what is the relation between `text` and `value` in the documentation? 

Interesting case is: if there is an original value `12345` and the user deletes `3` and gets `1245`. What will the IDE send as a message?

Reply:
The `value` field covers exactly this case. IDE sends an edit removing the character between positions 2 and 3, and the new value (in pseudo json)
```json
{ 
  "edit" : {
    "range" : [2,3],
    "text" : "",
    "kind" : {
      "expressionId" : "uuid",
      "value" : "1245"
    }
}
```

Reply:
Making the IDE aware of _identifier change_ is a great intention. However when the engine applies the change it also needs to rebuild the `IR` (and Truffle nodes). It is not entirely clear how _correct_ the new IR has to be, but the current implementation needs the new `IR` element to have proper location - see [ChangesetBuilder](https://github.com/enso-org/enso/blob/c72a6582bc62ced2e7da200688728774d661fb7b/engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala#L90) - otherwise subsequent changes may not find the right element to change.

Right now the beginning of the `IR` is the same as the initial edit location. That's no longer going to be the case if the edit is done _in middle_ of an identifier....

Maybe it is not a problem if `expressionId` is used more extensively. Just make sure subsequent changes to the literal work.